Add support for distributing as a .zip file in addition to .tar.gz, increment plugin version number.

shawnallen85 opened the pull request:
Added outputting a distribution as a .zip file since AWS Lambda expects the payload to be in that format.

shawnallen85 committed:
Distribute as .zip file

Added support for distributing as a .zip file. FaaS offerings expect the payload to be a .zip rather than .tar.gz. Default output is .tar.gz for backwards compatibility.

shawnallen85 committed:
Increment version to 0.2.19

andyczerwonka commented:
Hi @shawnallen85, @karlmartens will be back from vacation mid-next week, at which point he will be able to look at this PR.

shawnallen85 commented:
Any update on this? :-)

karlmartens commented on src/main/java/net/karlmartens/dotnet/Distribution.java:
```diff
@@ -118,20 +120,42 @@ private void archive(File file) throws IOException {
-            for (String includedFile : scanner.getIncludedFiles()) {
-                LOGGER.debug("Adding file {}", includedFile);
-                File f = source.resolve(includedFile).toFile();
+        switch(getExtension().getDistributeAs()) {
```
This switch statement should be over an enumeration and the case implementations extracted into methods. 

```
OutputStream out = new FileOutputStream(targetFile, false);
switch(getExtension().getDistributeAs()) {
  case TAR:
    produceTar(scanner, out);
    break;

  case ZIP:
    produceZip(scanner, out);
    break;

  default:
    throw new IllegalStateException();
}
```
The above implementation would lead to a significant amount of duplicated code and based on the DRY principle a better implementation would be to introduce an abstraction for the archive and as we scan the included files add files to it. This might look like:

```
with (Archive archive = Archive.getInstance(getExtension().getDistributeAs(), targetFile)) {
    for (String includedFile : scanner.getIncludedFiles()) {
        LOGGER.debug("Adding file {}", includedFile);
        File f = source.resolve(includedFile).toFile();
        archive.add(f);
    }
}
```

shawnallen85 committed:
Migrated compression to Apache Compression Commons

Doesn't line up completely with suggestions, but does bring in a layer of abstraction for archiving and compressing. I don't have any unit tests for this to make sure things work as expected. Thoughts?

karlmartens committed:
Merge branch 'master' into master
